### PR TITLE
clientv3: fix balancer update address bug

### LINKED ISSUE #7189: TestDialSetEndpoints: context deadline exceeded
via https://semaphoreci.com/coreos/etcd/branches/master/builds/1443
```
--- FAIL: TestDialSetEndpoints (3.22s)
	dial_test.go:57: context deadline exceeded
```

Possibly related to recent balancer changes. /cc @fanminshi

### PR DESCRIPTION
**Pull request by fanminshi**
FIX #7189

### CODE REVIEW

File: clientv3/balancer.go [hunk 1]
```diff
@@ -124,7 +124,18 @@ func (b *simpleBalancer) updateAddrs(eps []string) {
 		addrs = append(addrs, grpc.Address{Addr: getHost(eps[i])})
 	}
 	b.addrs = addrs
-	b.notifyCh <- addrs
+	// only notify grpc if the existing connection is down.
```
Comment (heyitsanthony) [inline clientv3/balancer.go]: really wordy-- how about `// updating notifyCh can trigger new connections, but balancer only expects new connections if all connections are down`?

Comment (fanminshi) [inline clientv3/balancer.go]: will change.

**Comment by fanminshi** [comment]
all fixed. PTAL @heyitsanthony

File: clientv3/integration/dial_test.go [hunk 2]
```diff
@@ -27,33 +27,56 @@ import (
 
 // TestDialSetEndpoints ensures SetEndpoints can replace unavailable endpoints with available ones.
 func TestDialSetEndpoints(t *testing.T) {
+	dialSetEndpoints(t, func(eps []string, clus *integration.ClusterV3) *clientv3.Client {
```
Comment (heyitsanthony) [inline clientv3/integration/dial_test.go]: Probably simpler to just have a flag and update with two endpoints regardless:

```diff
--- a/clientv3/integration/dial_test.go
+++ b/clientv3/integration/dial_test.go
@@ -25,8 +25,16 @@ import (
        "golang.org/x/net/context"
 )
 
-// TestDialSetEndpoints ensures SetEndpoints can replace unavailable endpoints with available ones.
-func TestDialSetEndpoints(t *testing.T) {
+func TestDialSetEndpointsBeforeFail(t *testing.T) {
+       testDialSetEndpoints(t, true)
+}
+
+func TestDialSetEndpointsAfterFail(t *testing.T) {
+       testDialSetEndpoints(t, false)
+}
+
+// testDialSetEndpoints ensures SetEndpoints can replace unavailable endpoints with available ones.
+func testDialSetEndpoints(t *testing.T, setBefore bool) {
        defer testutil.AfterTest(t)
        clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
        defer clus.Terminate(t)
@@ -45,12 +53,15 @@ func TestDialSetEndpoints(t *testing.T) {
        }
        defer cli.Close()
 
+       if setBefore {
+               cli.SetEndpoints(eps[toKill%3], eps[(toKill+1)%3])
+       }
        // make a dead node
        clus.Members[toKill].Stop(t)
        clus.WaitLeader(t)
-
-       // update client with available endpoints
-       cli.SetEndpoints(eps[(toKill+1)%3])
+       if !setBefore {
+               cli.SetEndpoints(eps[toKill%3], eps[(toKill+1)%3])
+       }
```

**Comment by fanminshi** [comment]
All fixed. PTAL cc\ @heyitsanthony

File: clientv3/integration/dial_test.go [hunk 3]
```diff
@@ -45,13 +54,18 @@ func TestDialSetEndpoints(t *testing.T) {
...
 
+	if setBefore {
+		cli.SetEndpoints(eps[toKill%3], eps[(toKill+1)%3])
+	}
 	// make a dead node
 	clus.Members[toKill].Stop(t)
 	clus.WaitLeader(t)
 
 	// update client with available endpoints
 	cli.SetEndpoints(eps[(toKill+1)%3])
```
Comment (heyitsanthony) [inline clientv3/integration/dial_test.go]: can remove this line?

**Comment by heyitsanthony** [comment]
lgtm after removing that extra SetEndpoints. Thanks!

File: clientv3/integration/dial_test.go (hunk 3 above)
Comment (fanminshi) [inline clientv3/integration/dial_test.go]: yep, good catch!

**Comment by fanminshi** [comment]
merge once CI passes

**Pull request merged**